packages/pds/src/sequencer/sequencer-leader.ts

```diff
@@ -0,0 +1,127 @@
+import { DisconnectError } from '@atproto/xrpc-server'
+import { jitter, wait } from '@atproto/common'
+import { Leader } from '../db/leader'
+import { seqLogger as log } from '../logger'
+import Database from '../db'
+
+export const SEQUENCER_LEADER_ID = 1100
+
+export class SequencerLeader {
+  leader: Leader
+
+  destroyed = false
+  polling = false
+  queued = false
+  lastSeq: number
+
+  constructor(public db: Database) {
+    const lockId = this.db.getSchemaAdvisoryLockId(SEQUENCER_LEADER_ID)
+    this.leader = new Leader(lockId, this.db)
+  }
+
+  nextSeqVal(): number {
+    this.lastSeq++
+    return this.lastSeq
+  }
+
+  async run() {
+    if (this.db.dialect === 'sqlite') return
+
+    const res = await this.db.db
+      .selectFrom('repo_seq')
+      .select('seq')
+      .orderBy('seq', 'desc')
+      .limit(1)
+      .executeTakeFirst()
+    this.lastSeq = res?.seq ?? 0
+
+    while (!this.destroyed) {
+      try {
+        const { ran } = await this.leader.run(async ({ signal }) => {
+          const seqListener = () => {
+            if (this.polling) {
+              this.queued = true
+            } else {
+              this.polling = true
+              this.pollDb()
+            }
+          }
+          this.db.channels.new_repo_event.addListener('message', seqListener)
+          await new Promise<void>((resolve) => {
+            signal.addEventListener('abort', () => {
+              this.db.channels.new_repo_event.removeListener(
+                'message',
+                seqListener,
+              )
+              resolve()
+            })
+          })
+        })
+        if (ran && !this.destroyed) {
+          throw new Error(
+            'Sequencer leader completed, but should be persistent',
+          )
+        }
+      } catch (err) {
+        log.error({ err }, 'sequence leader errored')
+      } finally {
+        if (!this.destroyed) {
+          await wait(1000 + jitter(500))
+        }
+      }
+    }
+  }
+
+  async pollDb() {
+    if (this.destroyed) {
+      this.polling = false
+      this.queued = false
+      return
+    }
+
+    try {
+      await this.sequenceOutgoing()
+    } catch (err) {
+      log.error({ err }, 'sequencer leader failed to sequence batch')
+    } finally {
+      // check if we should continue polling
+      if (this.queued === false) {
+        this.polling = false
+      } else {
+        this.queued = false
+        this.pollDb()
+      }
+    }
+  }
+
+  async sequenceOutgoing() {
+    const unsequenced = await this.getUnsequenced()
+    for (const row of unsequenced) {
+      await this.db.db
+        .updateTable('repo_seq')
+        .set({ seq: this.nextSeqVal() })
+        .where('id', '=', row.id)
+        .execute()
+      await this.db.notify('outgoing_repo_seq')
+    }
```

Comment:
With a little elbow grease this could become a batch update, which we might find pretty worthwhile if we expect `unsequenced` to often contain multiple rows.  Something like:
```sql
UPDATE repo_seq
SET 
  seq = vals.seq
FROM (
  VALUES
    (1, nextSeqVal()),
    (2, nextSeqVal())
) AS vals (id, seq)
WHERE repo_seq.id = vals.id
```